pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,633 @@
+	defer utilruntime.HandleCrash()
+
+	defer c.leaseQueue.ShutDown()
+	klog.Infof("Workers: %d", workers)
+	for i := 0; i < workers; i++ {
+		klog.Infof("Starting worker")
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func NewGenerator(ctx context.Context, config *rest.Config, leaseInformer coordinationv1.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	var cancel context.CancelFunc
+
+	callbacks := leaderelection.LeaderCallbacks{
+		OnStartedLeading: func(ctx context.Context) {
+			var genCtx context.Context
+			controller, err := NewController(leaseInformer, leaseCandidateInformer, leaseClient, leaseCandidateClient)
+			if err != nil {
+				panic(err)
+			}
+			genCtx, cancel = context.WithCancel(ctx)
+			controller.Sync(genCtx)
+			controller.Run(genCtx, 2)
+		},
+		OnStoppedLeading: func() {
+			cancel()
+		},
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	rl, err := resourcelock.NewFromKubeconfig(
+		"leases",
+		"kube-system",
+		"leader-election-controller",
+		resourcelock.ResourceLockConfig{
+			Identity: hostname + "_" + string(uuid.NewUUID()),
+		},
+		config,
+		10,
+	)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: Configurable deadlines
+	go leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: 10 * time.Second,
+		RenewDeadline: 5 * time.Second,
+		RetryPeriod:   2 * time.Second,
+		Callbacks:     callbacks,
+		Name:          "leader-election-controller",
+	})
+	return nil, nil
+}
+
+func NewController(leaseInformer coordinationv1.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	klog.Infof("NewController")
+	c := &Controller{
+		leaseInformer:          leaseInformer,
+		leaseCandidateInformer: leaseCandidateInformer,
+		leaseQueue:             workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[string](), workqueue.TypedRateLimitingQueueConfig[string]{Name: controllerName}),
+		leaseClient:            leaseClient,
+		leaseCandidateClient:   leaseCandidateClient,
+
+		// TODO: What to do if the size limit is reached?
+		// realistically we'd have a lot less than 1000 leases that all need to be updated at the same time.
+		// Is there any harm to naively spinning a goroutine for each election?
+		// I don't estimate the load on the apiserver to be that much
+		// Need to write producer code to handle? It is unsafe to block on send.
+		electionCh:               make(chan election, 1000),
+		electionCandidateTracker: make(map[leaderLeaseID]*candidateStatus),
```

Comment:
why not run in the workers the election loop directly and use the workqueue to guarantee there are no duplicate process? we can bound the parallelism with the number of workers, and based on the time to process an election make an estimation on max throughput